### ISSUE TITLE
BBoxes from TREK-150: confused as to format

### ISSUE THREAD

**Issue by vineetparikh** [matteo-dunnhofer/TREK-150-toolkit#5]
Hi there, thanks so much for the super helpful dataset!

I wanted to see what would happen if I prompted SAM with the boxes from TREK-150 (i.e. is it possible to get a mask of the object within each of the boxes), so following the current documentation/README I set up TREK-150 and visualized the boxes and corresponding masks for P03_02 and the P03_02-56 sequence. However, it seems like although the boxes are listed as xywh (where x and y are the top left coordinates), when converting to xyxy the corresponding boxes don't seem to match to objects when visualized (instead they're just parts of the environment, drift to objects, or are somewhat ambiguous with multiple objects). Any idea as to what's going on?

Here are a couple of screenshots:
![image](https://github.com/matteo-dunnhofer/TREK-150-toolkit/assets/13814267/4b519d6a-5812-43ea-87d0-d1f5d9ed4a33)
![image](https://github.com/matteo-dunnhofer/TREK-150-toolkit/assets/13814267/d46991cf-fd10-4265-80f3-23c4697e9db2)

My initial suspicion is that I wasn't handling the coordinate system correctly, but after a few permutations, I'm unsure if there's a clear "missing piece" here. My initial code is a bit sequential as I'm figuring out how this works before using utilities like `np.loadtxt`.

My code looks like the following (mostly taken from the SAM demo and reading through the TREK-150 readme):
```
from tracker_modules.bbox_to_mask import SamBBoxToMask
import argparse
from decord import VideoReader

import os
import numpy as np
import torch
import matplotlib.pyplot as plt
import cv2
from tqdm import tqdm

def show_mask(mask, ax, random_color=False):
    if random_color:
        color = np.concatenate([np.random.random(3), np.array([0.6])], axis=0)
    else:
        color = np.array([30/255, 144/255, 255/255, 0.6])
    h, w = mask.shape[-2:]
    mask_image = mask.reshape(h, w, 1) * color.reshape(1, 1, -1)
    ax.imshow(mask_image)
    
def show_points(coords, labels, ax, marker_size=375):
    pos_points = coords[labels==1]
    neg_points = coords[labels==0]
    ax.scatter(pos_points[:, 0], pos_points[:, 1], color='green', marker='*', s=marker_size, edgecolor='white', linewidth=1.25)
    ax.scatter(neg_points[:, 0], neg_points[:, 1], color='red', marker='*', s=marker_size, edgecolor='white', linewidth=1.25)   
    
def show_box(box, ax):
    x0, y0 = box[0], box[1]
    w, h = box[2] - box[0], box[3] - box[1]
    ax.add_patch(plt.Rectangle((x0, y0), w, h, edgecolor='green', facecolor=(0,0,0,0), lw=2))    

# These are symlinked
video_path = "assets/epic-kitchens-55-torrent/videos/train/P03/P03_02.MP4"
trek_folder_path = "assets/epic-kitchens-trek-150/P03/P03_02/P03_02-56"

 # Load video
video = VideoReader(video_path)

# Load trek zip folder
frame_file = os.path.join(trek_folder_path, "frames.txt")
gt_file = os.path.join(trek_folder_path, "groundtruth_rect.txt")
frame_id_mask_lst = []
bbox_to_mask = SamBBoxToMask()

# Making bboxes and masks
with open(frame_file, "r") as ff, open(gt_file, "r") as gf:
    frame_ids = ff.readlines()
    gts = gf.readlines()
    frame_ids = [int(frame.strip()) for frame in frame_ids]
    gt_bboxes = [list(map(int, gt.strip().split(","))) for gt in gts]
    frames = video.get_batch(frame_ids).asnumpy()
    gt_bboxes = [np.array([bbox[0], bbox[1], bbox[0]+bbox[2], bbox[1]+bbox[3]]) if bbox !=[-1,-1,-1,-1] else None for bbox in gt_bboxes
    for i, (frame_id, frame, bbox) in enumerate(tqdm(zip(frame_ids, frames, gt_bboxes))):
        if bbox is None:
            continue
        mask = bbox_to_mask.create_masks_from_bbox(frame, bbox)
        frame_id_mask_lst.append((frame_id, frame, bbox, mask))

# Visualization code
for frame_id, frame, bbox, mask in frame_id_mask_lst:
    print("Frame id: ", frame_id)
    plt.figure(figsize=(10,10))
    plt.imshow(frame)
    show_mask(mask[0], plt.gca())
    show_box(bbox, plt.gca())
    plt.axis('off')
    plt.show()
    plt.close()
```

**Comment by matteo-dunnhofer** [comment]
Hi @vineetparikh 

I noticed that you are decoding frames with `VideoReader`. We used `ffmpeg` to decode frames. I suspect that the two methods decode the frames differently (we experienced something similar in other projects), and so there might be a different number of frames per video, ultimately leading to a wrong synchronisation of frames and bounding boxes. Can you try to run your code on the frames we release in TREK-150? You can download them using the toolkit.

**Comment by vineetparikh** [comment]
^ will do and will report back if this is the case!

**Comment by vineetparikh** [comment]
looks like different ffmpeg versions were the case (`decord` packages a different ffmpeg version). Pulling the dataset via the toolkit, I'm getting more reasonable bboxes for the same objects at least for this sequence, so the frame IDs need to be extracted in exactly the same way.

![image](https://github.com/matteo-dunnhofer/TREK-150-toolkit/assets/13814267/c11a2845-c314-4000-a952-cf7a0c495cd7)

Thanks for the clarification! I'm going to close this issue but hopefully this will be helpful to someone else who has a similar problem moving forward.
